core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -554,25 +549,31 @@ private object RTS {
                     val io = curIo.asInstanceOf[IO.Supervise[E, Any]]
 
                     curIo = enterSupervision *>
-                      io.value.ensuring(exitSupervision(io.error))
+                      io.value.ensuring(exitSupervision)
 
                   case IO.Tags.Terminate =>
                     val io = curIo.asInstanceOf[IO.Terminate[E, Any]]
 
-                    val cause = io.cause
+                    val causes = io.causes
 
                     val finalizer = interruptStack
 
                     if (finalizer eq null) {
                       // No finalizers, simply produce error:
                       curIo = null
-                      result = ExitResult.Terminated(cause)
+                      val allCauses = status.get match {
+                        case Executing(Some(ts), _, _)            => causes ++ ts
+                        case AsyncRegion(Some(ts), _, _, _, _, _) => causes ++ ts
+                        case _                                    => causes
+                      }
+
+                      result = ExitResult.Terminated(allCauses)
 
                       // Report the termination cause to the supervisor:
-                      rts.submit(rts.unsafeRun(unhandled(cause)))
+                      rts.submit(rts.unsafeRun(unhandled(allCauses)))
```

Comment:
Ditto for this. It's too easy to miss calling this or to call it in a different way. Best to handle it uniformly.